### PR TITLE
CBL-704: Fix: Include user generated websocket closure codes

### PR DESCRIPTION
**Pull request by jayahariv**
* transient and permanent status codes by user was previously handled in platform side, now moving it to lite core. 
* include the transient code maybe_transient method. 

linking closed PR: https://github.com/couchbase/couchbase-lite-core/pull/923

### CODE REVIEW

File: C/include/c4Socket.h [hunk 1]
```diff
@@ -42,7 +42,9 @@ extern "C" {
         kWebSocketCloseCantFulfill      = 1011, // Can't fulfill request due to "unexpected condition"
         kWebSocketCloseTLSFailure       = 1015, // Never sent, only received
 
-        kWebSocketCloseFirstAvailable   = 4000, // First unregistered code for freeform use
+        kWebSocketCloseUserTransient    = 4001, // transient websocket closure error initiated by user
```
Comment (jayahariv) [inline C/include/c4Socket.h]:
```suggestion
        kWebSocketCloseUserTransient    = 4001, // transient websocket closure error by user
```

File: C/include/c4Socket.h [hunk 2]
```diff
@@ -42,7 +42,9 @@ extern "C" {
...
+        kWebSocketCloseUserPermanent    = 4002, // permanent websocket closure error initiated by user
```
Comment (jayahariv) [inline C/include/c4Socket.h]:
```suggestion
        kWebSocketCloseUserPermanent    = 4002, // permanent websocket closure error by user
```

File: C/include/c4Socket.h [hunk 3]
```diff
@@ -42,7 +42,9 @@ extern "C" {
         kWebSocketCloseCantFulfill      = 1011, // Can't fulfill request due to "unexpected condition"
         kWebSocketCloseTLSFailure       = 1015, // Never sent, only received
 
-        kWebSocketCloseFirstAvailable   = 4000, // First unregistered code for freeform use
+        kWebSocketCloseUserTransient    = 4001, // transient websocket closure error by user
```
Comment (jayahariv) [inline C/include/c4Socket.h]:
```suggestion
        kWebSocketCloseUserTransient    = 4001, // transient websocket closure by user
```

File: C/include/c4Socket.h [hunk 4]
```diff
@@ -42,7 +42,9 @@ extern "C" {
...
+        kWebSocketCloseUserPermanent    = 4002, // permanent websocket closure error by user
```
Comment (jayahariv) [inline C/include/c4Socket.h]:
```suggestion
        kWebSocketCloseUserPermanent    = 4002, // permanent websocket closure by user
```

**Comment by snej** [comment]
I'm not clear on what these codes are for. "closure by user" is kind of vague. @pasin, could you explain?

**Comment by borrrden** [comment]
These are used in the enterprise `MessageEndpoint` implementation to indicate that an exception occurred during the user implementation of their transport.  We don't know what kinds of exceptions they will face and the ones we define may not make sense, so we created two custom ones to use.  One (4002) is for an exception that should not be retried, and one (4001) is for an exception that should trigger a retry according to the retry logic (similar to a connection time out, etc).

**Comment by pasin** [comment]
Yes, it is for what @borrrden has explained.

File: C/include/c4Socket.h [hunk 5]
```diff
@@ -42,7 +42,9 @@ extern "C" {
         kWebSocketCloseCantFulfill      = 1011, // Can't fulfill request due to "unexpected condition"
         kWebSocketCloseTLSFailure       = 1015, // Never sent, only received
 
-        kWebSocketCloseFirstAvailable   = 4000, // First unregistered code for freeform use
+        kWebSocketCloseUserTransient    = 4001, // transient websocket closure by user
```
Comment (jayahariv) [inline C/include/c4Socket.h]:
```suggestion
        kWebSocketCloseUserTransient    = 4001, // any transient message endpoint web socket closure by user
```

Comment (jayahariv) [inline C/include/c4Socket.h]:
```suggestion
        kWebSocketCloseUserTransient    = 4001, // any transient message endpoint websocket closure by user
```

File: C/include/c4Socket.h [hunk 6]
```diff
@@ -42,7 +42,9 @@ extern "C" {
...
+        kWebSocketCloseUserPermanent    = 4002, // permanent websocket closure by user
```
Comment (jayahariv) [inline C/include/c4Socket.h]:
```suggestion
        kWebSocketCloseUserPermanent    = 4002, // any permanent message endpoint websocket closure by user
```

**Comment by snej** [comment]
Do you mean an error, like some problem with their transport, or an unexpected problem? Is there a way the app code could provide us an error code to use? Otherwise the error comes out the other end (back from the Replicator) as 4001/4002 code that may not mean anything to them.

**Comment by borrrden** [comment]
That's the way things have been now, but we've been saving the actual message from the error to go along with it so that the user has an idea of what is going on.  We could take an error code from them, but what if it overlaps with one of our numerous other error codes?

File: C/c4Base.cc [hunk 7]
```diff
@@ -245,6 +246,7 @@ bool c4error_mayBeTransient(C4Error err) C4API {
         websocket::kCodeGoingAway,
         websocket::kCodeAbnormal,
         websocket::kCodeUnexpectedCondition,
+        kWebSocketCloseUserTransient,
```
Comment (snej) [inline C/c4Base.cc]: The error codes in WebSocketInterface.hh should be kept in sync with the ones in c4Socket.h.

File: C/include/c4Socket.h [hunk 8]
```diff
@@ -42,7 +42,9 @@ extern "C" {
         kWebSocketCloseCantFulfill      = 1011, // Can't fulfill request due to "unexpected condition"
         kWebSocketCloseTLSFailure       = 1015, // Never sent, only received
 
-        kWebSocketCloseFirstAvailable   = 4000, // First unregistered code for freeform use
+        kWebSocketCloseUserTransient    = 4001, // any transient message endpoint websocket closure by user
```
Comment (snej) [inline C/include/c4Socket.h]: How about `// App-defined transient error`, and `// App-defined permanent error`

File: C/include/c4Socket.h [hunk 9]
```diff
@@ -42,7 +42,9 @@ extern "C" {
...
+        kWebSocketCloseUserPermanent    = 4002, // any permanent message endpoint websocket closure by user
+        kWebSocketCloseFirstAvailable   = 4003, // First unregistered code for freeform use
```
Comment (snej) [inline C/include/c4Socket.h]: Bump this up to 5000 so it won't be affected if we have to add more errors above it. (Also check that none of the platforms are using codes in the 4000 range.)

File: C/c4Base.cc (hunk 7 above)
Comment (jayahariv) [reply C/c4Base.cc]: https://github.com/couchbaselabs/BLIP-Cpp/pull/20

**Pull request merged**